src/libstd/io/mod.rs

```diff
@@ -0,0 +1,928 @@
+        // Create a shim which translates a Writer to a fmt::Writer and saves
+        // off I/O errors. instead of discarding them
+        struct Adaptor<'a, T: ?Sized + 'a> {
+            inner: &'a mut T,
+            error: Result<()>,
+        }
+
+        impl<'a, T: Write + ?Sized> fmt::Writer for Adaptor<'a, T> {
+            fn write_str(&mut self, s: &str) -> fmt::Result {
+                match self.inner.write_all(s.as_bytes()) {
+                    Ok(()) => Ok(()),
+                    Err(e) => {
+                        self.error = Err(e);
+                        Err(fmt::Error)
+                    }
+                }
+            }
+        }
+
+        let mut output = Adaptor { inner: self, error: Ok(()) };
+        match fmt::write(&mut output, fmt) {
+            Ok(()) => Ok(()),
+            Err(..) => output.error
+        }
+    }
+}
+
+/// Extension methods for all instances of `Write`, typically imported through
+/// `std::io::prelude::*`.
+pub trait WriteExt: Write + Sized {
+    /// Create a "by reference" adaptor for this instance of `Write`.
+    ///
+    /// The returned adaptor also implements `Write` and will simply borrow this
+    /// current writer.
+    fn by_ref(&mut self) -> &mut Self { self }
+
+    /// Creates a new writer which will write all data to both this writer and
+    /// another writer.
+    ///
+    /// All data written to the returned writer will both be written to `self`
+    /// as well as `other`. Note that the error semantics of the current
+    /// implementation do not precisely track where errors happen. For example
+    /// an error on the second call to `write` will not report that the first
+    /// call to `write` succeeded.
+    fn broadcast<W: Write>(self, other: W) -> Broadcast<Self, W> {
+        Broadcast { first: self, second: other }
+    }
+}
+
+impl<T: Write> WriteExt for T {}
+
+/// An object implementing `Seek` internally has some form of cursor which can
+/// be moved within a stream of bytes.
+///
+/// The stream typically has a fixed size, allowing seeking relative to either
+/// end or the current offset.
+pub trait Seek {
+    /// Seek to an offset, in bytes, in a stream
+    ///
+    /// A seek beyond the end of a stream is allowed, but seeking before offset
+    /// 0 is an error.
+    ///
+    /// Seeking past the end of the stream does not modify the underlying
+    /// stream, but the next write may cause the previous data to be filled in
+    /// with a bit pattern.
+    ///
+    /// This method returns the new position within the stream if the seek
+    /// operation completed successfully.
+    ///
+    /// # Errors
+    ///
+    /// Seeking to a negative offset is considered an error
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64>;
+}
+
+/// Enumeration of possible methods to seek within an I/O object.
+#[derive(Copy, PartialEq, Eq, Clone, Debug)]
+pub enum SeekFrom {
+    /// Set the offset to the provided number of bytes.
+    Start(u64),
```

Comment:
@mahkoh you may want to bring up these concerns [on the RFC](https://github.com/rust-lang/rfcs/pull/576) as that's a part I don't personally feel comfortable changing in the implementation vs the RFC.